### PR TITLE
[第8組]Refactor/error handling

### PR DESCRIPTION
**Pull request by samlu00412**
## purpose
* 增加錯誤處理功能，以防專案出現錯誤時無法空置，導致專案崩潰
* 修改上週PR的部分建議
## changes

1. 於所有router中，以及其對應的services.py/depends.py新增error handling and log file generating function
2. 將config.py中的Constants class 分類
3. 新增logger功能

### CODE REVIEW

**Comment by Rogue8464** [comment]
這是一份十分好的PR，透過error_handlers中不同的exception充分的考慮到程式運行中可能發生的錯誤並將其用logger記錄下來，對於程式運行的穩定性很好

**Comment by Steven0811** [comment]
pricetracker.log, pricetracker.err應該放入.gitignore裡面
使用Template Method Pattern定義各種exception，但在應用上有一些問題，在你的程式碼中有很多類似下面的結構
```python
capture_exception(err)
return HTTPException(status_code=400, detail="Something went wrong while processing search keywords")
```
這樣的error handling並不好，首先是400 error代表客戶端錯誤，但是根據你的錯誤描述，這個問題出現在系統內部，另外這些HTTPException應該要被你定義在error_handlers的各種exception裡面，會直接使用HTTPException代表你在定義exception時並不完整

**Comment by haruuu1214** [comment]
小建議：logger 不用推到 git 上，db 也是。 
總體來說，這次 PR 的修改提升了系統的穩定性和可維護性，特別是日誌記錄與錯誤處理的優化非常實用。代碼整潔，邏輯清晰，沒有多餘的變動，表現出專業的開發水準，辛苦了！

**Comment by mlcp29th-hi** [comment]
I noticed a few areas in the current implementation where enhanced error handling could be beneficial to improve the robustness of our application. Here are some suggestions:

1. **Database Operations**:
   - In the logs, I see multiple `ROLLBACK` messages, which indicates that transactions are failing. It would be wise to implement structured error handling around database interactions. This could involve catching exceptions related to database operations (e.g., `SQLAlchemyError`) and logging specific messages about what operation failed. Additionally, consider implementing retries for transient errors.

2. **Request Handling**:
   - For external API calls (like those to Sentry), check the response status. If the response indicates an error (4xx or 5xx HTTP status codes), the application should handle those cases gracefully by logging the error and potentially retrying the request if appropriate. This will help ensure our application does not fail silently during external service outages.

3. **Initialization Failures**:
   - During application startup, various components are initialized (e.g., routers). If any of these initializations fail, we should capture those errors and prevent the application from continuing in an inconsistent state. We could log the specific initialization step that failed and provide fallback measures or clear error messages.

4. **User Authentication**:
   - When logging in users, if the username/password verification fails, we should provide informative logs with details on the failure without exposing sensitive information. Ensure we also handle edge cases like nonexistent users or database connectivity issues during login attempts.

5. **Job Scheduling**:
   - The log shows that jobs are added to a job store and scheduled. If job scheduling fails, it might lead to missed runs. Implement a mechanism to catch exceptions during job scheduling and log them appropriately. Depending on the job's criticality, you may want to trigger alerts or fallback mechanisms.

6. **Configuration and Environment Checks**:
   - It would be beneficial to check that all necessary configurations and environment variables are set correctly before the application fully initializes. This can help catch errors early and avoid unnecessary complications later in execution.

Please consider these recommendations for enhancing the error handling in our application. Improving our error management will increase reliability and provide a clearer understanding of problems when they arise.

**Comment by jxuanyang** [comment]
這個 PR 出色地加強了專案的錯誤處理機制，確保在問題發生時系統依然穩定，同時精確地改進了上次 PR 的建議，優化了代碼結構和可維護性

**Comment by Yi-Cheng-Wang** [comment]
## Advantages

1. **Unified Error-Handling Structure**:  
   By utilizing a base class (`ExceptionsBase`) from which all custom exceptions inherit, the error-handling logic is centralized, reducing code duplication and ensuring consistency across the application.  

2. **Detailed Error Information**:  
   Each exception class provides specific error messages, enabling developers to quickly identify and resolve issues. For example, `UserNotFoundError` and `InvalidModelError` deliver clear, actionable error details.  

3. **Sentry Integration**:  
   Incorporating Sentry into error handling allows for automatic exception capturing and reporting, facilitating effective error tracking and analysis.  

4. **HTTP Status Code Management**:  
   Each exception class is mapped to an appropriate HTTP status code, making it easier for API consumers to understand the result of their requests and handle errors effectively.  

## Disadvantages

1. **Increased Number of Exception Classes**:  
   A large number of exception classes can increase code complexity, especially for smaller projects, where such a design might be unnecessarily cumbersome.  

2. **Lack of Specific Context Information**:  
   While exception classes provide error messages, they may lack detailed contextual information, such as the exact location or state of the error, which could make debugging more challenging.  

3. **Redundant Error-Handling Logic**:  
   In some cases, similar error-handling logic might appear in multiple places, leading to code redundancy and higher maintenance overhead.  

4. **Unhandled Exceptions**:  
   Certain scenarios might involve unhandled exceptions, potentially causing the application to crash or behave unpredictably. Ensuring comprehensive coverage for all possible exceptions is critical.  

## Summary

This pull request establishes a solid foundation for error handling, offering effective mechanisms to capture and manage exceptions. However, as the project scales, optimizing the error-handling structure to reduce complexity and redundancy will be crucial. Enhancing the granularity of error details and ensuring all potential exceptions are addressed will further improve the system’s maintainability and robustness.

**Comment by Yuannzh1** [comment]
在這個PR中，新增了多個`try-except`塊來捕捉和處理異常，並添加了日誌記錄功能。

### 建議
1. **分解大型的`try-except`塊**：
    - 雖然目前的`try-except`塊大部分都很合理，但可以進一步分解一些較大的塊，以提高可讀性和可維護性。例如，在`authenticate_user_token`函數內，可以將JWT解碼和數據庫查詢分成兩個`try-except`塊。
    ```python
    try:
        logging.debug(f"Authenticating token: {token[:10]}...")
        payload = jwt.decode(token, Constants.Auth.KEY, algorithms=["HS256"])
    except ExpiredSignatureError as e:
        logging.error(f"Token expired: {e}")
        capture_exception(e)
        raise HTTPException(status_code=401, detail="Token expired")
    except JWTError as e:
        logging.error(f"Token invalid: {e}")
        capture_exception(e)
        raise HTTPException(status_code=401, detail="Invalid token")
    except Exception as e:
        logging.error(f"Failed to authenticate token: {e}")
        capture_exception(e)
        raise HTTPException(status_code=401, detail="Failed to authenticate token")

    try:
        user = database.query(User).filter(User.username == payload.get("sub")).first()
    except Exception as e:
        logging.error(f"Database query failed: {e}")
        capture_exception(e)
        raise HTTPException(status_code=500, detail="Database query failed")
    return user
    ```
2. **統一的異常處理**：
    - 可以考慮設置一個統一的異常處理函數，來處理和記錄異常，這樣可以減少重複代碼，並保證所有異常處理的一致性。

總的來說，這個PR在異常處理和日誌記錄方面做得很好，但可以通過進一步分解`try-except`塊和提供更具體的日誌消息來進一步改進。

File: backend/log/pricetracker.log
Comment (WANG-YongLE) [inline backend/log/pricetracker.log:1]: 這個不用丟上來

File: backend/src/configs.py [hunk 1]
```diff
@@ -6,21 +6,33 @@
...
+        PRICES_INFO_LINK = "https://opendata.ey.gov.tw/api/ConsumerProtection/NecessitiesPrice"
+        NEWS_LINK = "https://udn.com/api/more"
+    class Prompt:
+        GPT_SUMMARY_PROMPT = "你是一個新聞摘要生成機器人，請統整新聞中提及的影響及主要原因 (影響、原因各50個字，請以json格式回答 {'影響': '...', '原因': '...'})，並請確保返回有效 json 格式"
+        GPT_RELEVANCE_PROMPT = "你是一個關聯度評估機器人，請評估新聞標題是否與「民生用品的價格變化」相關，並給予'high'、'medium'、'low'評價。(僅需回答'high'、'medium'、'low'三個詞之一)"
+        GPT_EXTRACT_PROMPT = "你是一個關鍵字提取機器人，用戶將會輸入一段文字，表示其希望看見的新聞內容，請提取出用戶希望看見的關鍵字，請截取最重要的關鍵字即可，避免出現「新聞」、「資訊」等混淆搜尋引擎的字詞。(僅須回答關鍵字，若有多個關鍵字，請以空格分隔)"
+    class APIKey:
+        ANTHROPIC_TOKEN = os.getenv("ANTHROPIC_API_KEY", "")
+        OPENAI_TOKEN = os.getenv("OPENAI_API_KEY", "")
+        API_KEY = "xxx"
```
Comment (WANG-YongLE) [inline backend/src/configs.py:19]: API_KEY好像可以砍掉

**Comment by WANG-YongLE** [comment]
整體來說沒啥大問題，金主爸爸老強了

File: backend/log/pricetracker.err
Comment (dee9c-git) [inline backend/log/pricetracker.err:1]: Logs can be ignored

File: backend/log/pricetracker.log
Comment (dee9c-git) [inline backend/log/pricetracker.log:1]: Logs can be ignored

File: backend/src/auth/depends.py [hunk 2]
```diff
@@ -18,25 +20,48 @@ def verify(password, hashed_password):
...
+        raise HTTPException(status_code=401, detail="Failed to authenticate token")
     return database.query(User).filter(User.username == payload.get("sub")).first()
 
 def create_access_token(data, expires_delta=None):
     to_encode = data.copy()
     if expires_delta:
         expire = datetime.utcnow() + expires_delta
     else:
-        expire = datetime.utcnow() + timedelta(minutes=15)
+        expire = datetime.utcnow() + timedelta(minutes=Constants.Auth.DEFAULT_EXPIRED_TIME)
```
Comment (dee9c-git) [inline backend/src/auth/depends.py:59]: ~~datetime.utcnow()~~
https://blog.miguelgrinberg.com/post/it-s-time-for-a-change-datetime-utcnow-is-now-deprecated
Use this (I think) instead
```python
datetime.now(timezone.utc)
```

File: backend/src/configs.py (hunk 1 above)
Comment (dee9c-git) [inline backend/src/configs.py:19]: As mentioned by others, I don't think it's needed.

File: backend/src/crawler/udn_crawler.py [hunk 3]
```diff
@@ -107,32 +110,47 @@ def parse(self, url: str) -> News:
...
+                for paragraph in content_section.find_all("p")
+                if paragraph.text.strip() != "" and "▪" not in paragraph.text
+            ]
+
+            return News(
+                url=url,
+                title=title,
+                time=content_time,
+                content=" ".join(paragraphs)
+            )
```
Comment (dee9c-git) [inline backend/src/crawler/udn_crawler.py:136]: Large ```try``` statement

File: backend/src/error_handlers/user_exceptions.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    def get_status_code(self):
+        return Constants.User.ERROR_CODE
+    
+class UserNotFoundError(UserExceptionBase):
+    def get_messages(self):
+        return f"{self.username} not found."
+
+class UserAlreadyExistError(UserExceptionBase):
+    def get_messages(self):
+        return f"{self.username} Alerady exists, please login directly."
```
Comment (dee9c-git) [inline backend/src/error_handlers/user_exceptions.py:20]: Nitpicking: Alerady -> already.

File: backend/src/logger/logger_base.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        self.logger.debug(message.format())
+
+    def log_warning(self, message: LogMessageInterface):
+        self.logger.warning(message.format())
+
+    def log_error(self, message: LogMessageInterface):
+        self.logger.error(message.format())
+
+    def log_critical(self, message: LogMessageInterface):
+        self.logger.critical(message.format())
```
Comment (dee9c-git) [inline backend/src/logger/logger_base.py:76]: I'm not sure where this is used.

File: backend/src/util.py
Comment (dee9c-git) [inline backend/src/util.py:1]: Bruh, why the hell is this here?
This is just logger_base all over again.

**Review by New9c — COMMENTED** [review]
The logger is rather weird, with what I believe is unused utils.py and 
```log_info()``` and 4 other functions that aren't used?
Maybe I'm missing something, though.

File: backend/log/pricetracker.err
Comment (kevin0216) [inline backend/log/pricetracker.err:1]: It's strongly suggested that log file should not uploaded to GitHub repository, it should be ignored (Add *.log into .gitignore to ignore it)

File: backend/src/posts/routers.py [hunk 6]
```diff
@@ -47,53 +74,82 @@ def read_user_news(
 @router.post("/search_news")
 async def search_news(request: PromptRequest):
     """Input a prompt, and catch the data which AI finds."""
+    logging.debug(f"Accessed /api/v1/news/search_news: {request.prompt}")
+    if request.prompt == "":
+        raise NoPromptError()
```
Comment (kevin0216) [inline backend/src/posts/routers.py:79]: This exception is not being catched

File: backend/src/users/routers.py [hunk 7]
```diff
@@ -20,21 +24,33 @@ async def login_for_access_token(
...
+    if not user:
+        logging.debug(f"Failed to login: {form_data.username}")
+        raise HTTPException(status_code=400, detail="Incorrect username or password")
+    try:
+        access_token = create_access_token(
+            data={"sub": str(user.username)}, expires_delta=timedelta(minutes=30)
+        )
+    except Exception as e:
+        capture_exception(e)
+        return HTTPException(status_code=400, detail="Something went wrong while processing the request")
```
Comment (kevin0216) [inline backend/src/users/routers.py:38]: Logging is missing here (personally think this part should have logger)

File: backend/src/users/routers.py [hunk 8]
```diff
@@ -20,21 +24,33 @@ async def login_for_access_token(
...
+    logging.debug(f"Logged in: {form_data.username}")
     return {"access_token": access_token, "token_type": "bearer"}
 
 @router.post("/register")
 def create_new_user(user: UserAuthSchema, database: Session = Depends(session_opener)):
+    logging.debug(f"Creating user: {user.username}")
     hashed_password = password_context.hash(user.password)
     database_user = User(username=user.username, hashed_password=hashed_password)
     database.add(database_user)
     database.commit()
```
Comment (kevin0216) [inline backend/src/users/routers.py:48]: `Session.commit() `is possibly throwing exception, should add a try..except expression to handle exception

File: backend/src/users/routers.py [hunk 9]
```diff
@@ -20,21 +24,33 @@ async def login_for_access_token(
...
     database.refresh(database_user)
+    logging.debug(f"Created user: {user.username}")
     return database_user
 
 @router.get("/me")
 def read_users_name(user=Depends(authenticate_user_token)):
+    logging.debug(f"Accessed /api/v1/users/me: {user.username}")
     return {"username": user.username}
```
Comment (kevin0216) [inline backend/src/users/routers.py:56]: `return {"username": user.username}` is possibly throwing exception when user is None, should add a try..except expression to handle exception

**Pull request merged**